### PR TITLE
CI: Add a sanitizer CI job

### LINKED ISSUE #23116: Task: add address sanitizer CI run
With meson it's relatively simple to configure the build to use address sanitizer. We should leverage that to add a CI build that runs the tests with address sanitizer turned on.

c.f. https://github.com/scipy/scipy/pull/16447/files where support for doing this was added to scipy's meson build.

### PR DESCRIPTION
**Pull request by ngoldbaum**
Closes #23116

Adds a github actions job that runs the tests under both address sanitizer and undefined behavior sanitizer.

There are some undefined behavior errors when I run the tests locally so this will not fail the build if an error happens until those are fixed. I will be filing a separate issue about the undefined behavior reports.

### CODE REVIEW

**Comment by ngoldbaum** [comment]
Hmm, not sure why the new workflow isn't getting triggered here...

File: .github/workflows/linux_asan.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  PYTHON_VERSION: 3.11
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
+  cancel-in-progress: true
+
+permissions:
+  contents: read # to fetch code (actions/checkout)
+
+jobs:
```
Comment (r-devulap) [inline .github/workflows/linux_asan.yml:23]: I think your job needs a name here. Something like this: https://github.com/numpy/numpy/blob/2e668061db8967de50413f3e201e97edae4c4a75/.github/workflows/linux_meson.yml#L24

Comment (tylerjereddy) [reply .github/workflows/linux_asan.yml:23]: Yup this passes the lint locally:

```diff
--- a/.github/workflows/linux_asan.yml
+++ b/.github/workflows/linux_asan.yml
@@ -21,34 +21,35 @@ permissions:
  contents: read # to fetch code (actions/checkout)

jobs:
  {+sanitizer:+}
    if: "github.repository == 'numpy/numpy'"
    runs-on: ubuntu-latest
```

Can iterate locally with i.e., `act` tool:

```
Error: workflow is not valid. 'linux_asan.yml': yaml: unmarshal errors:
  line 24: cannot unmarshal !!str `github....` into model.Job
  line 25: cannot unmarshal !!str `ubuntu-...` into model.Job
  line 27: cannot unmarshal !!seq into model.Job
```

**Comment by charris** [comment]
`asan` isn't an informative name. Maybe spell it out, `test_with_memory_sanitizer` or some such.

**Comment by charris** [comment]
Thanks Nathan.

**Pull request merged**